app/src/main/java/org/oppia/android/app/customview/VerticalDashedLineView.kt

```diff
@@ -0,0 +1,67 @@
+package org.oppia.android.app.customview
+
+import android.content.Context
+import android.graphics.Canvas
+import android.graphics.DashPathEffect
+import android.graphics.Paint
+import android.util.AttributeSet
+import android.util.TypedValue
+import android.view.View
+import androidx.core.content.ContextCompat
+import org.oppia.android.R
+
+private const val DASH_GAP_IN_DP = 16
+private const val DASH_LENGTH_IN_DP = 32
+private const val DASH_THICKNESS_IN_DP = 6
+
+/**
+ * Custom view to represent chapter progress in [StoryActivity].
+ *
+ * Reference: https://stackoverflow.com/a/27054463
+ */
+class VerticalDashedLineView @JvmOverloads constructor(
+  context: Context,
+  attrs: AttributeSet? = null
+) : View(context, attrs) {
+  private val paint: Paint
+  private lateinit var canvas: Canvas
+
+  init {
+    val dashGap: Float = dpToPx(DASH_GAP_IN_DP)
+    val dashLength: Float = dpToPx(DASH_LENGTH_IN_DP)
+    val dashThickness: Float = dpToPx(DASH_THICKNESS_IN_DP)
+
+    paint = Paint()
+    paint.isAntiAlias = true
+    paint.color = ContextCompat.getColor(context, R.color.colorPrimary)
+    paint.style = Paint.Style.STROKE
+    paint.strokeCap = Paint.Cap.ROUND
+    paint.strokeWidth = dashThickness
+    paint.pathEffect = DashPathEffect(floatArrayOf(dashLength, dashGap), /* phase= */ 0F)
+  }
+
+  fun setColor(colorId: Int) {
+    paint.color = colorId
+    drawLine()
+  }
+
+  override fun onDraw(canvas: Canvas) {
+    this.canvas = canvas
```

Comment:
Is it safe to cache the canvas like this?

Reply:
Actually I am not fully sure if this is safe or not and therefore I have updated the code to remove the global `canvas` completely.

Reply:
Thanks--this seems much better and more in-line with what I'd expect of a custom view.